### DISCUSSION TITLE
Content based recommendation system

### DISCUSSION THREAD

**Discussion by arsaboo** [beetbox/beets#4689] [Ideas]
Now that we have the basic infrastructure (ratings, track/album characteristics, popularity information), I think we should be able to create a basic recommendation system for smart playlists. The idea would be to predict or rate tracks that users may like based on their prior ratings or consumption habits. 

This would be extremely useful to create a "smart" playlist. Wondering if there is an appetite for such a feature...

**Comment by RollingStar** [comment]
Have you seen calliope?

https://discourse.beets.io/t/calliope-antisocial-music-recommendations/1532

smart playlists is a holy grail for me, I'd love to see it (And see it grow over time). the paid streamers are repeatedly iterating on this. understandably, open source projects are a few years behind. spotify has been really impressive for years, but I prefer to buy direct from bandcamp rather than give the artists pennies.

You probably already know, but listen tracking (listenbrainz/lastfm) will be key here, imo. if you want smart playlists later, start tracking now.

**Reply by arsaboo** [reply]
@RollingStar I am tracking everything. All my music listening happens in Plex and my Plexsync plugin can track everything (plays, ratings, etc.). In addition, we can already get music attributes (rhythm, danceability, etc.) from Spotify. We have the basic information already - artists, year, genre, etc. I think creating a basic recommender should not be very difficult.

**Reply by gryphonmyers** [reply]
> @RollingStar I am tracking everything. All my music listening happens in Plex and my Plexsync plugin can track everything (plays, ratings, etc.). In addition, we can already get music attributes (rhythm, danceability, etc.) from Spotify. We have the basic information already - artists, year, genre, etc. I think creating a basic recommender should not be very difficult.

If you're already in the Plex ecosystem, can't you just use Plex's recommendations? That's what I'm doing. I think it would be cool if Beets did this but I'm not even sure how it would work since it's not handling playback.

**Reply by arsaboo** [reply]
@gryphonmyers I just added [Plexsonic](https://github.com/beetbox/beets/discussions/4794) that adds some GPT-based intelligence. In addition, I have a few private Jupyter notebooks that do content-based recommendations.

**Reply by gryphonmyers** [reply]
> @gryphonmyers I just added [Plexsonic](https://github.com/beetbox/beets/discussions/4794) that adds some GPT-based intelligence. In addition, I have a few private Jupyter notebooks that do content-based recommendations. 

Ok, but Plex already does recommendations (track playlists, album playlists, artist playlists, blending between tracks with Sonic Adventure, even chat gpt playlists, with Sonic Sage) so I'm a little confused about what your plugin does that Plex can't already do natively?

**Reply by arsaboo** [reply]
Sonic sage was the motivation for this implementation. Right now, plex does a very poor job of matching (see [here](https://www.reddit.com/r/plexamp/comments/13iv7w5/sonic_sage_not_so_hot/)). I have improved the matching process to get close to 100% match and even manual search. 

Other than that, the plugin allows importing curated playlists from Apple Music, Spotify, or other sources or even creating your own smart playlists. 

I have a huge library and use Plexamp exclusively for my music needs, but there are a lot of areas where Plex does a poor job (including discovery), which is where this plugin comes in. For example, I could now create playlists that are popular on Spotify from a particular artist. The primary motivation of the plugin was to bring all the other information (beyond MB that Plex already uses) to enable discovery.

**Reply by gryphonmyers** [reply]
> Sonic sage was the motivation for this implementation. Right now, plex does a very poor job of matching (see [here](https://www.reddit.com/r/plexamp/comments/13iv7w5/sonic_sage_not_so_hot/)). I have improved the matching process to get close to 100% match and even manual search.
> 
> Other than that, the plugin allows importing curated playlists from Apple Music, Spotify, or other sources or even creating your own smart playlists.
> 
> I have a huge library and use Plexamp exclusively for my music needs, but there are a lot of areas where Plex does a poor job (including discovery), which is where this plugin comes in. For example, I could now create playlists that are popular on Spotify from a particular artist. The primary motivation of the plugin was to bring all the other information (beyond MB that Plex already uses) to enable discovery.

I see - personally I don't use the ChatGPT feature, so  I was not aware of these shortcomings. I am more interested in the recommendation features that are based on sonic analysis, and those are metadata-independent. They also work pretty well! I listen to a lot of obscure music and it's able to help me discover interesting stuff hiding in my library.

**Reply by RollingStar** [reply]
I'm skeptical that chatGPT could provide useful recommendations. It's just going to be shuffling places like Allmusic and Rateyourmusic, if not hallucinating results entirely (like The Beatles? You'll love 100 gecs). Without insight into the content of the music itself (bpm, instruments, style/genre, lyrics), there's going to be a human in the loop - at the very least, annotating music for ML algorithms.

**Reply by arsaboo** [reply]
I don't think the objective is to take the human out of the loop. It is the same problem with using something like Spotify popularity information - I may not like a song that is very popular. ChatGPT gives some great recommendations (and I am sure the quality of these recommendations will only go up as more data/plugins become available) and is a great starting point already.  In addition, we have music attributes (from Spotify) or `xtractor` that we can use to improve the quality of recommendations. 

My workflow is as follows. I use these automated tools (ChatGPT or popularity-based playlists) to create a starting point. I then review them myself to see if I love those tracks and will rate them in Plex accordingly. Eventually, the plex rating is a function of music attributes and other meta data.

**Comment by gryphonmyers** [comment]
> Now that we have the basic infrastructure (ratings, track/album characteristics, popularity information), I think we should be able to create a basic recommendation system for smart playlists. The idea would be to predict or rate tracks that users may like based on their prior ratings or consumption habits.
> 
> This would be extremely useful to create a "smart" playlist. Wondering if there is an appetite for such a feature...

When you say we "have" ratings, track/album characteristics, popularity information, in what form do we have those? I was not aware of these features being in Beets. Do you mean through the use of your Plex plugin? What about for people who don't use Plex?

**Reply by arsaboo** [reply]
I am using a lot of sources to get these details. First, the official [Spotify](https://beets.readthedocs.io/en/latest/plugins/spotify.html) plugin provides `popularity` information and other audio features. Beyond Spotify, I am using a few community plugins to get the details - [Tidal](https://github.com/arsaboo/beets-tidal), [Youtube](https://github.com/arsaboo/beets-youtube), [JioSaavn](https://github.com/arsaboo/beets-jiosaavn), and [Plex](https://github.com/arsaboo/beets-plexsync). 

The overall idea is to get as much detail as possible about each track into Beets from diverse sources and use the information to create smart playlists in Plex (which is where most of my listening happens). The Plex plugin has many features to create, import, and modify playlists. Even if you are not using Plex, you can use the information collected to create smart playlists.

**Reply by gryphonmyers** [reply]
So it seems your vision here is to improve the quality of Plex's recommendations by sourcing metadata from more sources and feeding playlists into Plex rather then using the built-in Plex recommendation playlists. Thanks for explaining that - I think I have a clearer understanding now. I see this as three distinct concerns:

1) Augmenting the metadata in Beets. It sounds like you've done some work / thinking about this. I would say this makes sense as either core functionality or a standalone plugin that ONLY cares about gathering additional metadata into your Beets library. This may even be multiple plugins... e.g. https://github.com/beetbox/beets/issues/1719 for listenbrainz play counts, a separate plugin for spotify popularity, etc. 

2) Generating playlists from metadata. The quality of the generated playlists would likely improve with more metadata, but it would do its best with what it has available to it. _I think it's very important to avoid coupling this behavior to any specific metadata or plugin._ Also, this MAY make sense as additional functionality of https://beets.readthedocs.io/en/stable/plugins/smartplaylist.html rather than an entirely new plugin.

3) Syncing playlists generated by Beets to Plex, Musicbrainz, Spotify, wherever. This could be multiple plugins, pushing playlists to a variety of platforms. 

Does that sound about right?

**Reply by arsaboo** [reply]
Yeah, that is generally the idea. Plex recommendations are great, but I would like to bring in other details to improve the discovery aspect of Plex (or any other player for that matter). 

I completely agree that we should not be coupling come of these features and I have tried to create functions that can be used by other plugins. For example, importing playlist functions from other online sources (e.g., Spotify) can easily be used by other plugins. I would love any feedback/PRs to make it even more generalizable and better integrate with Beets.

**Reply by gryphonmyers** [reply]
> Yeah, that is generally the idea. Plex recommendations are great, but I would like to bring in other details to improve the discovery aspect of Plex (or any other player for that matter). 
> 
> I completely agree that we should not be coupling come of these features and I have tried to create functions that can be used by other plugins. For example, importing playlist functions from other online sources (e.g., Spotify) can easily be used by other plugins. I would love any feedback/PRs to make it even more generalizable and better integrate with Beets. 

I'm happy to give feedback, and possibly help with this. I am a little distracted with my wanted albums plugin at the moment.

 @sampsyo do you have any thoughts or suggestions about how these concerns should be addressed? Do you think any part of this belongs in core beets or in existing plugins (like my suggestion of enhancing existing smart playlists plugin vs making a new one)?
